source/exe/win32/service_base.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <processenv.h>
+#include <shellapi.h>
+#include <winsvc.h>
+
+#include <functional>
+#include <string>
+
+#include "envoy/event/signal.h"
+
+#include "exe/service_status.h"
+
+namespace Envoy {
+class ServiceBase {
+public:
+  virtual ~ServiceBase() = default;
+  /**
+   * Try to register Envoy as a service. For this we register `ServiceMain` and `Handler`
+   * @param ServiceBase instance of the service.
+   * @return false if Envoy can not be registered to run as service.
+   */
+  static bool TryRunAsService(ServiceBase& service);
+
+  ServiceBase(DWORD controlsAccepted = SERVICE_ACCEPT_STOP | SERVICE_ACCEPT_SHUTDOWN |
+                                       SERVICE_CONTROL_PRESHUTDOWN);
```

Comment:
This may prove problematic. Paraphrasing MSDN docs, preshutdown circumvents shutdown, the service process will be terminated upon its timeout before the server begins the shutdown phase (think in terms of pre-upgrade steps). It is highly discouraged owing to the additional shutdown delay.

But it might be especially problematic where envoy is deployed as the egress proxy. If services that envoy is proxying are required during shutdown, envoy's pre-shutdown would break the shutdown of those corresponding services.

We have a mechanism to put envoy in a failed state and let consumers route around it. But we have no mechanism to preserve envoy as the egress proxy, given the code above. preshutdown notice is mutually exclusive of shutdown notice.

Reply:
You are right, I verified it internally as well in another service that I own that we should not accept both SHUTDOWN and PRESHUTDOWN. removed in https://github.com/envoyproxy/envoy/pull/14352/commits/44ce57b0a84e1c80792ff756a5cf530a930eb15d